File: Balanced_Binary_Tree.py
'''
Created on 2014.7.25

Given a binary tree, determine if it is height-balanced.

For this problem, a height-balanced binary tree is defined 
as a binary tree in which the depth of the two subtrees of 
every node never differ by more than 1.


'''

# Definition for a  binary tree node

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # @param root, a tree node
    # @return a boolean
    def isBalanced(self, root):
        if root ==None:
            return True
        return self.detectBalance(root)[0]
    def detectBalance(self , root):
        if root.left == None and root.right == None:
            return True , 1
        leftDepth , rightDepth = 0 , 0
        leftFlag , rightFlag = True , True
        if root.left != None:
            leftFlag , leftDepth = self.detectBalance(root.left)
        if root.right != None:
            rightFlag , rightDepth = self.detectBalance(root.right)
        if leftFlag and rightFlag and abs(leftDepth - rightDepth) <= 1:
            return True , max(leftDepth , rightDepth) + 1
        else:
            return False , 0

File: Cplusplus/Balanced_Binary_Tree.cpp
/*
Leetcode : Balanced_Binary_Tree.cpp

Given a binary tree, determine if it is height-balanced.

For this problem, a height-balanced binary tree is defined as a

binary tree in which the depth of the two subtrees of every node never differ by more than 1.
*/


/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode *root) {
        vector<int> depth(2,0);
        return postOrder(root , depth);
    }


    bool postOrder(TreeNode *root , vector<int> &depth)
    {
        if(root == NULL)
        {
            return true;
        }
        vector<int> leftDepth(2,0);
        vector<int> rightDepth(2,0);
        bool lFlag = postOrder(root->left , leftDepth);
        bool rFlag = postOrder(root->right , rightDepth);
        if(lFlag && rFlag)
        {
            depth[0] = max(leftDepth[0] , leftDepth[1]) + 1;
            depth[1] = max(rightDepth[0] , rightDepth[1]) + 1;
            if(abs(depth[0] - depth[1]) < 2)
                return true;
            else
                return false;
        }
        else
            return false;
    }
};
